crucible-llvm/src/Lang/Crucible/LLVM/QQ.hs

```diff
@@ -0,0 +1,330 @@
+------------------------------------------------------------------------
+-- |
+-- Module           : Lang.Crucible.LLVM.QQ
+-- Description      : QuasiQuoters for a subset of LLVM assembly syntax
+-- Copyright        : (c) Galois, Inc 2019
+-- License          : BSD3
+-- Maintainer       : Rob Dockins <rdockins@galois.com>
+-- Stability        : provisional
+------------------------------------------------------------------------
+
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Lang.Crucible.LLVM.QQ
+ ( llvmType
+ , llvmDecl
+ , llvmOvr
+ ) where
+
+import Control.Monad (void)
+import qualified Data.Attoparsec.Text as AT
+import Data.Char
+import Data.Data
+import Data.Int
+import qualified Data.Text as T
+import qualified Text.LLVM.AST as L
+
+import Language.Haskell.TH
+import Language.Haskell.TH.Syntax
+import Language.Haskell.TH.Quote
+
+import qualified Data.Parameterized.Context as Ctx
+import           Lang.Crucible.Types
+import qualified Lang.Crucible.LLVM.Intrinsics.Common as IC
+import           Lang.Crucible.LLVM.Types
+
+data QQType
+  = QQVar String
+  | QQIntVar String
+  | QQSizeT
+  | QQPrim L.PrimType
+  | QQPtrTo QQType
+  | QQAlias L.Ident
+  | QQArray Int32 QQType
+  | QQFunTy QQType [QQType] Bool
+  | QQStruct [QQType]
+  | QQPackedStruct [QQType]
+  | QQVector Int32 QQType
+  | QQOpaque
+ deriving (Show, Eq, Ord, Data)
+
+data QQDeclare =
+  QQDeclare
+  { qqDecRet     :: QQType
+  , qqDecName    :: Either String L.Symbol
+  , qqDecArgs    :: [QQType]
+  , qqDecVarArgs :: Bool
+  }
+ deriving (Show, Eq, Ord, Data)
+
+parseIdent :: AT.Parser L.Ident
+parseIdent = L.Ident <$> (AT.char '%' *> AT.choice
+  [ T.unpack <$> AT.takeWhile1 isDigit
+  , (:) <$> AT.satisfy (AT.inClass "-a-zA-Z$._")
+        <*> (T.unpack <$> (AT.takeWhile (AT.inClass "-a-zA-Z$._0-9")))
+  ])
+
+
+parseSymbol :: AT.Parser L.Symbol
+parseSymbol = L.Symbol <$> (AT.char '@' *>
+  ( (:) <$> AT.satisfy (AT.inClass "-a-zA-Z$._")
+        <*> (T.unpack <$> (AT.takeWhile (AT.inClass "-a-zA-Z$._0-9")))
+  ))
+
+parseFloatType :: AT.Parser L.FloatType
```

Comment:
Like the types above, these parsers look similar-but-off from the content of https://github.com/elliottt/llvm-pretty/blob/master/src/Text/LLVM/Parser.hs. The difference here is the use of `attoparsec` vs `parsec`. Again, it probably makes sense to add a comment explaining why we can't reuse those parsers.

Reply:
Yeah, the parsers are quite similar, but differ slightly because of the parsing for quasiquoter metavariables; and, as you say, the parsing technology used.  Because of the quasiquoter variables, we can't just reuse the same parser as-is... but with some refactoring we might be able to unify them.  I'm not sure it's worth it.

@brianhuffman, any thoughts about this? It looks like you were mainly the author of the `llvm-pretty` parser in question.   What is it used for?

Reply:
That parser is used to implement the saw-script function `llvm_type : String -> LLVMType`, which lets you write LLVM types using LLVM type syntax. I thought it made more sense for the parser to go in the llvm-pretty package than in the saw-script package. If you can think of a way it could be generalized to make it more useful, then maybe we should do that.

Reply:
Well, actually, that's another place where having additional variable-binding forms might be useful.  What do you think about pushing the `QQType` and related types into `llvm-pretty` along with the parsers.  Then you could have the `llvm_type` function grab variables from the local context as well.